File: 13-workflows/python-sdk/create_workflow.py
"""
Cost Katana Workflows: Create and Execute Workflows (Python)

Orchestrate multi-step AI operations with automatic orchestration.

Run: python 17-workflows/python-sdk/create_workflow.py
"""

import os
import time
import requests
from typing import Dict, List, Any

API_BASE = "https://api.costkatana.com/api"
API_KEY = os.getenv("COST_KATANA_API_KEY")

def create_workflow_template(template: Dict[str, Any]) -> Dict:
    """Create a workflow template"""
    
    response = requests.post(
        f"{API_BASE}/workflows/templates",
        json=template,
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type": "application/json"
        }
    )
    
    response.raise_for_status()
    data = response.json()["data"]
    
    print(f"✅ Workflow template created:")
    print(f"   ID: {data['id']}")
    print(f"   Name: {data['name']}")
    print(f"   Steps: {len(data['steps'])}")
    
    return data

def execute_workflow(template_id: str, variables: Dict[str, Any]) -> Dict:
    """Execute a workflow"""
    
    response = requests.post(
        f"{API_BASE}/workflows/templates/{template_id}/execute",
        json={"variables": variables},
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type": "application/json"
        }
    )
    
    response.raise_for_status()
    data = response.json()["data"]
    
    print(f"✅ Workflow execution started:")
    print(f"   Execution ID: {data['executionId']}")
    print(f"   Status: {data['status']}")
    
    return data

def get_execution_status(execution_id: str) -> Dict:
    """Get workflow execution status"""
    
    response = requests.get(
        f"{API_BASE}/workflows/executions/{execution_id}",
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    return response.json()["data"]

def wait_for_completion(execution_id: str, max_wait=60) -> Dict:
    """Wait for workflow to complete"""
    
    start_time = time.time()
    
    while time.time() - start_time < max_wait:
        status = get_execution_status(execution_id)
        
        if status["status"] == "completed":
            print("\n✅ Workflow completed successfully!")
            return status
        
        if status["status"] == "failed":
            print("\n❌ Workflow failed")
            return status
        
        elapsed = int(time.time() - start_time)
        print(f"⏳ Status: {status['status']}... ({elapsed}s)")
        time.sleep(2)
    
    raise TimeoutError("Workflow execution timeout")

def display_results(execution: Dict):
    """Display workflow results"""
    
    print("\n📊 Workflow Results:")
    print(f"   Duration: {execution['duration']}ms")
    print(f"   Total Cost: ${execution.get('totalCost', 0):.4f}")
    print(f"   Status: {execution['status']}")
    
    print("\n   Steps:")
    for i, step in enumerate(execution.get('steps', []), 1):
        print(f"   {i}. {step['name']}")
        print(f"      Status: {step['status']}")
        print(f"      Duration: {step['duration']}ms")
        print(f"      Cost: ${step.get('cost', 0):.4f}")
        if step.get('output'):
            output = str(step['output'])[:100]
            print(f"      Output: {output}...")

def document_analysis_example():
    """Example: Document analysis workflow"""
    
    print("🥷 Cost Katana Workflows: Document Analysis (Python)\n")
    
    # 1. Create template
    print("1️⃣ Creating workflow template...")
    template = {
        "name": "Document Analysis Workflow",
        "description": "Extract, analyze, and summarize documents",
        "steps": [
            {
                "id": "step_1",
                "name": "extract_text",
                "type": "ai",
                "model": "gpt-4",
                "prompt": "Extract all important information from: {{document}}",
                "output": "extracted_text"
            },
            {
                "id": "step_2",
                "name": "extract_entities",
                "type": "ai",
                "model": "gpt-4",
                "prompt": "Extract entities from: {{extracted_text}}",
                "dependsOn": ["step_1"],
                "output": "entities"
            },
            {
                "id": "step_3",
                "name": "summarize",
                "type": "ai",
                "model": "gpt-3.5-turbo",
                "prompt": "Summarize: {{extracted_text}}",
                "dependsOn": ["step_1"],
                "output": "summary"
            }
        ],
        "variables": [
            {
                "name": "document",
                "type": "string",
                "required": True
            }
        ],
        "config": {
            "maxRetries": 3,
            "retryDelay": 1000,
            "timeout": 60000
        }
    }
    
    created_template = create_workflow_template(template)
    
    # 2. Execute workflow
    print("\n2️⃣ Executing workflow...")
    execution = execute_workflow(created_template["id"], {
        "document": "John Smith works at Acme Corporation in New York. Founded in 1990, specializing in technology."
    })
    
    # 3. Wait for completion
    print("\n3️⃣ Waiting for completion...")
    result = wait_for_completion(execution["executionId"])
    
    # 4. Display results
    display_results(result)
    
    print("\n💡 Workflow Features:")
    print("   ✅ Multi-step orchestration")
    print("   ✅ Automatic dependencies")
    print("   ✅ Cost tracking")
    print("   ✅ Error handling")

def main():
    if not API_KEY:
        print("❌ COST_KATANA_API_KEY required")
        return
    
    try:
        document_analysis_example()
        print("\n✅ Workflow example complete!")
    except requests.exceptions.HTTPError as e:
        print(f"❌ Error: {e.response.json() if e.response else e}")

if __name__ == "__main__":
    main()

File: 13-workflows/npm-package/create-workflow.ts
/**
 * Cost Katana Workflows: Create and Execute Workflows
 * 
 * Orchestrate multi-step AI operations with automatic dependency
 * management, retries, and error handling.
 * 
 * Run: npx ts-node 17-workflows/npm-package/create-workflow.ts
 */

import axios from 'axios';

const API_BASE = 'https://api.costkatana.com/api';
const API_KEY = process.env.COST_KATANA_API_KEY;

interface WorkflowStep {
  id: string;
  name: string;
  type: 'ai' | 'function' | 'condition';
  model?: string;
  prompt?: string;
  dependsOn?: string[];
  output: string;
  parallel?: boolean;
}

interface WorkflowTemplate {
  name: string;
  description: string;
  steps: WorkflowStep[];
  variables: Array<{
    name: string;
    type: string;
    required: boolean;
    description?: string;
  }>;
  config?: {
    maxRetries?: number;
    retryDelay?: number;
    timeout?: number;
    failureHandling?: 'stop' | 'continue';
  };
}

/**
 * Create a workflow template
 */
async function createWorkflowTemplate(template: WorkflowTemplate) {
  try {
    const response = await axios.post(
      `${API_BASE}/agent-trace/templates`,
      template,
      {
        headers: {
          'Authorization': `Bearer ${API_KEY}`,
          'Content-Type': 'application/json'
        }
      }
    );

    console.log('✅ Workflow template created:');
    console.log(`   ID: ${response.data.data.id}`);
    console.log(`   Name: ${response.data.data.name}`);
    console.log(`   Steps: ${response.data.data.steps.length}`);

    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed to create template:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Execute a workflow
 */
async function executeWorkflow(templateId: string, variables: Record<string, any>) {
  try {
    const response = await axios.post(
      `${API_BASE}/agent-trace/templates/${templateId}/execute`,
      { variables },
      {
        headers: {
          'Authorization': `Bearer ${API_KEY}`,
          'Content-Type': 'application/json'
        }
      }
    );

    console.log('✅ Workflow execution started:');
    console.log(`   Execution ID: ${response.data.data.executionId}`);
    console.log(`   Status: ${response.data.data.status}`);

    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed to execute workflow:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Get workflow execution status
 */
async function getExecutionStatus(executionId: string) {
  try {
    const response = await axios.get(
      `${API_BASE}/agent-trace/executions/${executionId}`,
      {
        headers: {
          'Authorization': `Bearer ${API_KEY}`
        }
      }
    );

    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed to get execution status:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Wait for workflow to complete
 */
async function waitForCompletion(executionId: string, maxWaitTime = 60000): Promise<any> {
  const startTime = Date.now();
  const pollInterval = 2000;

  while (Date.now() - startTime < maxWaitTime) {
    const status = await getExecutionStatus(executionId);

    if (status.status === 'completed') {
      console.log('\n✅ Workflow completed successfully!');
      return status;
    }

    if (status.status === 'failed') {
      console.log('\n❌ Workflow failed');
      return status;
    }

    console.log(`⏳ Status: ${status.status}... (${Math.round((Date.now() - startTime) / 1000)}s)`);
    await new Promise(resolve => setTimeout(resolve, pollInterval));
  }

  throw new Error('Workflow execution timeout');
}

/**
 * Display workflow results
 */
function displayResults(execution: any) {
  console.log('\n📊 Workflow Results:');
  console.log(`   Duration: ${execution.duration}ms`);
  console.log(`   Total Cost: $${execution.totalCost?.toFixed(4)}`);
  console.log(`   Status: ${execution.status}`);

  console.log('\n   Steps:');
  execution.steps?.forEach((step: any, index: number) => {
    console.log(`   ${index + 1}. ${step.name}`);
    console.log(`      Status: ${step.status}`);
    console.log(`      Duration: ${step.duration}ms`);
    console.log(`      Cost: $${step.cost?.toFixed(4)}`);
    if (step.output) {
      const outputPreview = typeof step.output === 'string' 
        ? step.output.substring(0, 100) 
        : JSON.stringify(step.output).substring(0, 100);
      console.log(`      Output: ${outputPreview}...`);
    }
  });

  if (execution.outputs) {
    console.log('\n   Final Outputs:');
    Object.entries(execution.outputs).forEach(([key, value]) => {
      console.log(`   ${key}: ${typeof value === 'string' ? value.substring(0, 100) : JSON.stringify(value).substring(0, 100)}...`);
    });
  }
}

/**
 * Example: Document Analysis Workflow
 */
async function documentAnalysisExample() {
  console.log('🥷 Cost Katana Workflows: Document Analysis\n');

  // 1. Create workflow template
  console.log('1️⃣ Creating workflow template...');
  const template: WorkflowTemplate = {
    name: 'Document Analysis Workflow',
    description: 'Extract, analyze, and summarize documents',
    steps: [
      {
        id: 'step_1',
        name: 'extract_text',
        type: 'ai',
        model: 'gpt-4',
        prompt: 'Extract all important information from this document: {{document}}',
        output: 'extracted_text'
      },
      {
        id: 'step_2',
        name: 'extract_entities',
        type: 'ai',
        model: 'gpt-4',
        prompt: 'From this text, extract all entities (people, organizations, locations, dates): {{extracted_text}}',
        dependsOn: ['step_1'],
        output: 'entities'
      },
      {
        id: 'step_3',
        name: 'summarize',
        type: 'ai',
        model: 'gpt-3.5-turbo',
        prompt: 'Create a concise summary of this text: {{extracted_text}}',
        dependsOn: ['step_1'],
        output: 'summary'
      }
    ],
    variables: [
      {
        name: 'document',
        type: 'string',
        required: true,
        description: 'The document content to analyze'
      }
    ],
    config: {
      maxRetries: 3,
      retryDelay: 1000,
      timeout: 60000,
      failureHandling: 'continue'
    }
  };

  const createdTemplate = await createWorkflowTemplate(template);

  // 2. Execute workflow
  console.log('\n2️⃣ Executing workflow...');
  const execution = await executeWorkflow(createdTemplate.id, {
    document: 'John Smith works at Acme Corporation in New York. The company was founded in 1990 and specializes in technology solutions. John has been with the company for 5 years and serves as the VP of Engineering.'
  });

  // 3. Wait for completion
  console.log('\n3️⃣ Waiting for workflow to complete...');
  const result = await waitForCompletion(execution.executionId);

  // 4. Display results
  displayResults(result);

  console.log('\n💡 Workflow Features Used:');
  console.log('   ✅ Multi-step orchestration');
  console.log('   ✅ Automatic dependency management');
  console.log('   ✅ Variable substitution');
  console.log('   ✅ Cost tracking per step');
  console.log('   ✅ Automatic retries');
}

/**
 * Main function
 */
async function main() {
  if (!API_KEY) {
    console.error('❌ COST_KATANA_API_KEY environment variable required');
    process.exit(1);
  }

  try {
    await documentAnalysisExample();

    console.log('\n✅ Workflow example complete!');
    console.log('\n📚 Next steps:');
    console.log('   - Try parallel execution (see parallel-workflows.ts)');
    console.log('   - Add conditional branching (see conditional-workflows.ts)');
    console.log('   - Implement error handling (see error-handling.ts)');

  } catch (error) {
    console.error('\n❌ Example failed:', error);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export {
  createWorkflowTemplate,
  executeWorkflow,
  getExecutionStatus,
  waitForCompletion,
  displayResults
};

